### PR TITLE
raftstore-v2: support apply snapshot

### PR DESCRIPTION
**Pull request by bufferflies**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/12842

What's Changed:

```commit-message
aftstore-v2: support apply snapshot
This PR implementation apply snapshot relative changes.
- add snap worker for multi-rocksdb


```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/async_io/write.rs [hunk 1]
```diff
@@ -800,7 +828,12 @@ where
         let mut callback_time = 0f64;
         if notify {
             for (region_id, (peer_id, ready_number)) in &self.batch.readies {
-                self.notifier.notify(*region_id, *peer_id, *ready_number);
+                let mut need_scheduled = false;
```
Comment (BusyJay) [inline components/raftstore/src/store/async_io/write.rs:831]: It's very strange to let writer tell peer to reschedule, it feels like though `after_write_hook` is an arbitrary closure, but writer know it has to be snapshot. Whether need to reschedule should be decided by the peer itself, writer only needs to persist change, call the hook and notify peer it's persisted.

File: components/raftstore-v2/src/operation/ready/async_writer.rs [hunk 2]
```diff
@@ -124,11 +127,15 @@ impl<EK: KvEngine, ER: RaftEngine> AsyncWriter<EK, ER> {
         }
 
         let mut raft_messages = vec![];
+        let mut has_snapshot = false;
         // There must be a match in `self.unpersisted_readies`.
         loop {
             let Some(v) = self.unpersisted_readies.pop_front() else {
                 panic!("{:?} ready number not found {}", logger.list(), ready_number);
             };
+            if v.has_snapshot {
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/async_writer.rs:136]: You mean `hash_snapshot |= v.has_snapshot`?

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 3]
```diff
@@ -383,9 +399,20 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             error!(self.logger, "peer id not matched"; "persisted_peer_id" => peer_id, "persisted_number" => ready_number);
             return;
         }
-        let persisted_message = self
-            .async_writer
-            .on_persisted(ctx, ready_number, &self.logger);
+        let (persisted_message, has_snapshot) =
+            self.async_writer
+                .on_persisted(ctx, ready_number, &self.logger);
+        if has_snapshot {
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/mod.rs:405]: Only when `persisted_index == first_index && has_snapshot` should it assume the snapshot is applied.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 4]
```diff
@@ -313,4 +316,71 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
...
+        snap: &Snapshot,
+        task: &mut WriteTask<EK, ER>,
+        snap_mgr: TabletSnapManager,
+        tablet_factory: Arc<dyn TabletFactory<EK>>,
+    ) -> Result<()> {
+        let region_id = self.get_region_id();
+        let peer_id = self.peer().get_id();
+        info!(self.logger(),
+            "begin to apply snapshot";
+            "region_id"=> region_id,
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:342]: These labels are included in `logger`.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 5]
```diff
@@ -313,4 +316,71 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
...
+            "region_id" => region_id,
+            "peer_id" => peer_id,
+            "state" => ?self.entry_storage().apply_state(),
+        );
+
+        let key = TabletSnapKey::new(region_id, peer_id, last_term, last_index);
+        let mut path = snap_mgr.get_recv_tablet_path(&key);
+
+        let hook =
+            move |region_id: u64| tablet_factory.load_tablet(path.as_path(), region_id, last_index);
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:382]: region_id can be captured directly, no need to pass it again.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 6]
```diff
@@ -313,4 +316,71 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
...
+        task.add_after_write_hook(Some(Box::new(hook)));
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:383]: Should add comment to explain why it's OK to load tablet after write (instead of before).

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 7]
```diff
@@ -383,9 +399,20 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+            self.storage_mut().after_applied_snapshot();
+            let suffix = self.storage().raft_state().last_index;
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/mod.rs:407]: suffix should be first index.

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 8]
```diff
@@ -499,11 +526,23 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+        ready: &mut Ready,
+        write_task: &mut WriteTask<EK, ER>,
+        ctx: &mut StoreContext<EK, ER, T>,
+    ) {
         let prev_raft_state = self.entry_storage().raft_state().clone();
         let ever_persisted = self.ever_persisted();
 
-        // TODO: handle snapshot
+        if !ready.snapshot().is_empty() {
+            let _ = self.apply_snapshot(
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/mod.rs:539]: Why ignore the returned value?

File: components/raftstore-v2/src/operation/ready/async_writer.rs (hunk 2 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/async_writer.rs:136]: yes

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 4 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs:342]: fixed

File: components/raftstore/src/store/async_io/write.rs (hunk 1 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs]: fixed

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 3 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: the first_index is the truncate_index+1 and the persisted_index is the truncate_index , so the condition should be persisted_index == first_index-1

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 4 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs:342]: fixed

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 5 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs]: fixed

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 7 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fixed

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 8 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fixed by log the error info.

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 6 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs:379]: fixed

File: components/raftstore/src/store/async_io/write.rs [hunk 9]
```diff
@@ -195,9 +199,20 @@ where
...
             trackers: vec![],
+            after_write_hook: None,
+            has_snapshot: false,
         }
     }
 
+    pub fn on_after_write_to_kv_db(&mut self) {
+        if self.after_write_hook.is_some() {
+            let hook = std::mem::take(&mut self.after_write_hook).unwrap();
+            if let Err(e) = hook() {
```
Comment (BusyJay) [inline components/raftstore/src/store/async_io/write.rs:210]: The hook should not return any error. async writer has no idea what's the proper way to handle.

File: components/raftstore/src/store/async_io/write.rs [hunk 10]
```diff
@@ -499,9 +514,10 @@ where
     }
 
     fn after_write_to_kv_db(&mut self, metrics: &StoreWriteMetrics) {
-        if metrics.waterfall_metrics {
-            let now = std::time::Instant::now();
-            for task in &self.tasks {
+        let now = std::time::Instant::now();
+        for task in &mut self.tasks {
```
Comment (BusyJay) [inline components/raftstore/src/store/async_io/write.rs:518]: In most cases, there will be no callback. Better use some way to avoid looping.

File: components/raftstore/src/store/async_io/write.rs [hunk 11]
```diff
@@ -706,10 +722,8 @@ where
                 write_kv_time = duration_to_sec(now.saturating_elapsed());
                 STORE_WRITE_KVDB_DURATION_HISTOGRAM.observe(write_kv_time);
             }
-
-            self.batch.after_write_to_kv_db(&self.metrics);
         }
-
+        self.batch.after_write_to_kv_db(&self.metrics);
```
Comment (BusyJay) [inline components/raftstore/src/store/async_io/write.rs:726]: If there is no write to kv db, then it's wrong to call `after_write_to_kv_db`.

File: components/raftstore/src/store/entry_storage.rs [hunk 12]
```diff
@@ -958,6 +958,26 @@ impl<EK: KvEngine, ER: RaftEngine> EntryStorage<EK, ER> {
         }
     }
 
+    #[inline]
+    pub fn get_truncate_index(&self) -> u64 {
```
Comment (BusyJay) [inline components/raftstore/src/store/entry_storage.rs:962]: Bad style in Rust.

```suggestion
    pub fn truncate_index(&self) -> u64 {
```

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 13]
```diff
@@ -499,11 +529,25 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
 
-        // TODO: handle snapshot
+        if !ready.snapshot().is_empty() {
+            if let Err(e) = self.apply_snapshot(
+                ready.snapshot(),
+                write_task,
+                ctx.snap_mgr.clone(),
+                ctx.tablet_factory.clone(),
+            ) {
+                error!(self.logger(),"failed to apply snapshot";"error"=>?e)
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fix

File: components/raftstore/src/store/async_io/write.rs (hunk 9 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs]: fix

File: components/raftstore/src/store/async_io/write.rs (hunk 10 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs]: yes, I will collect these hook  to  one vector on calling `add_write_task`

File: components/raftstore/src/store/async_io/write.rs (hunk 11 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs]: Understand it, I wiil execute it alone not in after_write_to_kv_db.

File: components/raftstore/src/store/entry_storage.rs (hunk 12 above)
Comment (bufferflies) [reply components/raftstore/src/store/entry_storage.rs]: fixed

File: components/raftstore/src/store/async_io/write.rs [hunk 14]
```diff
@@ -511,6 +520,12 @@ where
         }
     }
 
+    fn on_after_writer_to_kv_db(&mut self) {
```
Comment (BusyJay) [inline components/raftstore/src/store/async_io/write.rs:523]: In v2, all states are written to raft db, so the hook should be called after finishing all writes.

File: components/raftstore/src/store/entry_storage.rs [hunk 15]
```diff
@@ -958,6 +958,26 @@ impl<EK: KvEngine, ER: RaftEngine> EntryStorage<EK, ER> {
         }
     }
 
+    #[inline]
+    pub fn truncate_index(&self) -> u64 {
```
Comment (BusyJay) [inline components/raftstore/src/store/entry_storage.rs:962]: I think there is already `truncated_index` and `truncated_term` methods.

File: components/raftstore/src/store/snap.rs [hunk 16]
```diff
@@ -150,6 +150,10 @@ impl SnapKey {
             snap,
         ))
     }
+
+    pub fn get_snapshot_recv_path(&self) -> String {
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:154]: Why the change? V2 should not use v1's snapshot.

File: components/raftstore/src/store/snap.rs [hunk 17]
```diff
@@ -1956,7 +1968,12 @@ impl TabletSnapManager {
     }
 
     pub fn get_tablet_checkpointer_path(&self, key: &TabletSnapKey) -> PathBuf {
-        let prefix = format!("{}_{}", SNAP_GEN_PREFIX, key);
+        let prefix = key.get_gen_suffix();
+        PathBuf::from(&self.base).join(prefix)
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:1972]: What's wrong with the original implementation?

Comment (tonyxuqqi) [reply components/raftstore/src/store/snap.rs:1972]: @BusyJay I think both way are fine. In fact the new way is better IMO as it potentially eliminates the duplicate code of "format!("{}_{}", SNAP_GEN_PREFIX, key)"  in case somewhere else needs the path.

File: components/raftstore/src/store/snap.rs [hunk 18]
```diff
@@ -1910,6 +1914,14 @@ impl TabletSnapKey {
...
         TabletSnapKey::new(region_id, to_peer, term, index)
     }
+
+    pub fn get_gen_suffix(&self) -> String {
+        format!("{}_{}", SNAP_GEN_PREFIX, self)
+    }
+
+    pub fn get_recv_suffix(&self) -> String {
+        format!("{}_{}", SNAP_REV_PREFIX, self)
+    }
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/snap.rs:1924]: these function should be inline

File: components/raftstore/src/store/snap.rs (hunk 16 above)
Comment (tonyxuqqi) [reply components/raftstore/src/store/snap.rs:154]: Also it looks that get_snapshot_recv_path is not called at all.

File: components/raftstore/src/store/entry_storage.rs (hunk 15 above)
Comment (tonyxuqqi) [reply components/raftstore/src/store/entry_storage.rs:962]: get_truncate_index 
get_truncate_term

File: components/raftstore/src/store/snap.rs (hunk 17 above)
Comment (bufferflies) [reply components/raftstore/src/store/snap.rs:1960]: fixed， I has remove the extra modify.

File: components/raftstore/src/store/snap.rs (hunk 18 above)
Comment (bufferflies) [reply components/raftstore/src/store/snap.rs]: fixed

File: components/raftstore/src/store/entry_storage.rs (hunk 15 above)
Comment (bufferflies) [reply components/raftstore/src/store/entry_storage.rs]: yes，I have found the similar func

File: components/raftstore/src/store/async_io/write.rs (hunk 14 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs:523]: yes, I wiil call it after write raft db

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/batch/store.rs [hunk 19]
```diff
@@ -465,7 +471,7 @@ impl<EK: KvEngine, ER: RaftEngine> StoreRouter<EK, ER> {
     ) -> std::result::Result<(), TrySendError<Box<RaftMessage>>> {
         let id = msg.get_region_id();
         let peer_msg = PeerMsg::RaftMessage(msg);
-        let store_msg = match self.try_send(id, peer_msg) {
+        let store_msg = match self.router.try_send(id, peer_msg) {
```
Comment (tabokie) [inline components/raftstore-v2/src/batch/store.rs:474]: Why change this?

File: components/raftstore-v2/tests/integrations/test_conf_change.rs [hunk 20]
```diff
@@ -1,13 +1,90 @@
...
+    raft_serverpb::RaftMessage,
+};
 use raft::prelude::ConfChangeType;
+use raftstore_v2::router::PeerMsg;
 use tikv_util::store::new_learner_peer;
 
 use crate::cluster::Cluster;
 
+#[test]
+fn test_add_learner() {
```
Comment (tabokie) [inline components/raftstore-v2/tests/integrations/test_conf_change.rs:16]: Put this after `test_simple_change`.

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 21]
```diff
@@ -383,17 +399,38 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
         self.raft_group_mut().on_persist_ready(persisted_number);
         let persisted_index = self.raft_group().raft.raft_log.persisted;
+        let first_index = self.storage().first_index().unwrap() - 1;
+        /// The apply snapshot process order would be:
+        /// - Get the snapshot from the ready
+        /// - Wait for async write to load this tablet
+        /// In this step, the snapshot has loaded finish, but some apply state
+        /// need to update.
+        if first_index == persisted_index && has_snapshot {
+            let region_id = self.storage().region_id();
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/ready/mod.rs:421]: What's the difference with `self.region_id()`?

File: components/raftstore/src/store/async_io/write.rs (hunk 14 above)
Comment (tabokie) [inline components/raftstore/src/store/async_io/write.rs:523]:
```suggestion
    fn after_write_all(&mut self) {
```

File: components/raftstore/src/store/async_io/write.rs [hunk 22]
```diff
@@ -169,12 +169,15 @@ where
     ready_number: u64,
     pub send_time: Instant,
     pub raft_wb: Option<ER::LogBatch>,
+    // after_write_hook will be called after write to db.
+    pub after_write_hook: Option<Box<dyn FnOnce() + Send>>,
```
Comment (tabokie) [inline components/raftstore/src/store/async_io/write.rs:173]:
```suggestion
    // called after writing to kvdb and raftdb.
    pub persisted_cb: Option<Box<dyn FnOnce() + Send>>,
```

File: components/raftstore-v2/src/batch/store.rs (hunk 19 above)
Comment (bufferflies) [reply components/raftstore-v2/src/batch/store.rs:474]: It can be analized by vs code

File: components/raftstore-v2/tests/integrations/test_conf_change.rs (hunk 20 above)
Comment (bufferflies) [reply components/raftstore-v2/tests/integrations/test_conf_change.rs]: fixed

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 21 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fixed by using region_id().

File: components/raftstore/src/store/async_io/write.rs (hunk 14 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs]: fixed

File: components/raftstore/src/store/async_io/write.rs (hunk 22 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/write.rs]: fixed

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 23]
```diff
@@ -329,7 +345,7 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         let ready_number = ready.number();
         let mut write_task = WriteTask::new(self.region_id(), self.peer_id(), ready_number);
         self.storage_mut()
-            .handle_raft_ready(&mut ready, &mut write_task);
+            .handle_raft_ready(&mut ready, &mut write_task, ctx);
```
Comment (nolouch) [inline components/raftstore-v2/src/operation/ready/mod.rs:348]: order better be: (ctx, ready, write_task)

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 24]
```diff
@@ -383,17 +399,38 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
             }
         }
+
         let persisted_number = self.async_writer.persisted_number();
         self.raft_group_mut().on_persist_ready(persisted_number);
         let persisted_index = self.raft_group().raft.raft_log.persisted;
+        let first_index = self.storage().first_index().unwrap() - 1;
+        /// The apply snapshot process order would be:
+        /// - Get the snapshot from the ready
+        /// - Wait for async write to load this tablet
```
Comment (nolouch) [inline components/raftstore-v2/src/operation/ready/mod.rs:417]: async writer

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 25]
```diff
@@ -313,4 +316,69 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
         *snap_state = SnapState::Generated(snap);
         true
     }
+
+    pub fn after_applied_snapshot(&mut self) {
```
Comment (nolouch) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:320]: on_applied_snapshot

File: components/raftstore-v2/src/raft/storage.rs [hunk 26]
```diff
@@ -15,7 +15,10 @@ use raft::{
     eraftpb::{ConfState, Entry, Snapshot},
     GetEntriesContext, RaftState, INVALID_ID,
 };
-use raftstore::store::{util, EntryStorage, ReadTask, RAFT_INIT_LOG_INDEX, RAFT_INIT_LOG_TERM};
+use raftstore::store::{
+    util, util::new_empty_snapshot, EntryStorage, ReadTask, WriteTask, RAFT_INIT_LOG_INDEX,
```
Comment (nolouch) [inline components/raftstore-v2/src/raft/storage.rs:19]: It only used in test, should be `use` in test part.

File: components/raftstore-v2/tests/integrations/test_conf_change.rs [hunk 27]
```diff
@@ -47,6 +54,44 @@ fn test_simple_change() {
...
+    // trigger the learner execute persisted log.
+    std::thread::sleep(Duration::from_millis(1000));
+    cluster.dispatch(region_id, vec![]);
+
+    let meta = router1
+        .must_query_debug_info(region_id, Duration::from_secs(3))
+        .unwrap();
+    // the learner truncated index muse be equal the leader applied index
+    assert_eq!(match_index, meta.raft_apply.truncated_state.index);
+    assert!(meta.raft_apply.applied_index >= match_index);
```
Comment (nolouch) [inline components/raftstore-v2/tests/integrations/test_conf_change.rs:93]: Can we read `key` from the learner?

**Review by nolouch — COMMENTED** [review]
overall lgtm.

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 24 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fixed

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 23 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fixed

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 25 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs]: fixed

File: components/raftstore-v2/src/raft/storage.rs (hunk 26 above)
Comment (bufferflies) [reply components/raftstore-v2/src/raft/storage.rs]: fixed

File: components/raftstore-v2/tests/integrations/test_conf_change.rs (hunk 27 above)
Comment (bufferflies) [reply components/raftstore-v2/tests/integrations/test_conf_change.rs:95]: fixed

Comment (bufferflies) [reply components/raftstore-v2/tests/integrations/test_conf_change.rs:95]: fixed

**Comment by bufferflies** [comment]
PTAL @BusyJay again

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 28]
```diff
@@ -383,17 +399,38 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
         let persisted_number = self.async_writer.persisted_number();
         self.raft_group_mut().on_persist_ready(persisted_number);
         let persisted_index = self.raft_group().raft.raft_log.persisted;
+        let first_index = self.storage().first_index().unwrap() - 1;
+        /// The apply snapshot process order would be:
+        /// - Get the snapshot from the ready
+        /// - Wait for async writer to load this tablet
+        /// In this step, the snapshot has loaded finish, but some apply state
+        /// need to update.
+        if first_index == persisted_index && has_snapshot {
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/mod.rs:420]: Should be in snapshot.rs. And `+1` is better than `-1` as the latter is more likely to underflow.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 29]
```diff
@@ -313,4 +316,69 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
         *snap_state = SnapState::Generated(snap);
         true
     }
+
+    pub fn on_applied_snapshot(&mut self) {
+        let mut entry = self.entry_storage_mut();
+        let term = entry.truncated_term();
+        let index = entry.truncated_index();
+        entry.set_applied_term(term);
+        entry.set_last_term(term);
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:325]: This is actually a false assumption that last term will be the same as truncated term.

File: components/raftstore-v2/tests/integrations/test_conf_change.rs [hunk 30]
```diff
@@ -47,6 +55,50 @@ fn test_simple_change() {
         req.get_header().get_peer().get_id()
     );
 
+    let mut raft_msg = RaftMessage::default();
```
Comment (BusyJay) [inline components/raftstore-v2/tests/integrations/test_conf_change.rs:58]: Why need this? I think a message append should be sent immediately. Even if it's not sent, you can trigger the message by ticking a heartbeat.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 29 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs]: fixed

File: components/raftstore-v2/src/raft/storage.rs [hunk 31]
```diff
@@ -490,6 +497,58 @@ mod tests {
...
+            .unwrap();
+
+        // It can be set before load tablet.
+        assert_eq!(PeerState::Normal, s.region_state().get_state());
+        assert_eq!(10, s.entry_storage().truncated_index());
+        assert_eq!(1, s.entry_storage().truncated_term());
+        assert_eq!(10, s.entry_storage().raft_state().last_index);
+        // This index can't be set before load tablet.
+        assert_ne!(10, s.entry_storage().applied_index());
+        assert_ne!(1, s.entry_storage().last_term());
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply components/raftstore-v2/src/raft/storage.rs]: fixed

File: components/raftstore/src/store/snap.rs [hunk 32]
```diff
@@ -1959,6 +1959,11 @@ impl TabletSnapManager {
         let prefix = format!("{}_{}", SNAP_GEN_PREFIX, key);
         PathBuf::from(&self.base).join(prefix)
     }
+
+    pub fn get_recv_tablet_path(&self, key: &TabletSnapKey) -> PathBuf {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply components/raftstore/src/store/snap.rs]: fixed

File: components/raftstore-v2/tests/integrations/test_conf_change.rs (hunk 30 above)
Comment (bufferflies) [reply components/raftstore-v2/tests/integrations/test_conf_change.rs]: yes, I have use raft tick to trigger the snapshot

File: components/raftstore-v2/src/operation/ready/mod.rs (hunk 28 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: Sound good, I have move into `snapshot.rs` .

**Comment by nolouch** [comment]
ci failed @bufferflies

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 33]
```diff
@@ -115,6 +118,29 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             self.set_has_ready();
         }
     }
+
+    pub fn on_applied_snapshot<T: Transport>(&mut self, ctx: &mut StoreContext<EK, ER, T>) {
+        let persisted_index = self.raft_group().raft.raft_log.persisted;
+        let first_index = self.storage().entry_storage().truncated_index();
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:124]: This is not first_index. `first_index` has special meanings.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 34]
```diff
@@ -313,4 +339,64 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
...
+        self.entry_storage_mut().set_last_term(last_term);
+
+        let key = TabletSnapKey::new(region_id, peer_id, last_term, last_index);
+        let mut path = snap_mgr.final_recv_path(&key);
+        let logger = self.logger().clone();
+        // The snapshot require no additional processing such as ingest them to DB, but
+        // it should load it into the factory after it persisted.
+        let hook = move || {
+            if let Err(e) = tablet_factory.load_tablet(path.as_path(), region_id, last_index) {
+                error!(logger, "failed to load tablet";"path" => path.display(),"err" => ?e);
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:395]: It should panic instead of just logging. Fail to load tablet means data is corrupted.

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 35]
```diff
@@ -383,17 +399,28 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
         for msgs in persisted_message {
             for msg in msgs {
                 self.send_raft_message(ctx, msg);
             }
         }
+
         let persisted_number = self.async_writer.persisted_number();
         self.raft_group_mut().on_persist_ready(persisted_number);
         let persisted_index = self.raft_group().raft.raft_log.persisted;
+        let first_index = self.storage().first_index().unwrap() - 1;
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/mod.rs]: fixed

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 34 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/ready/snapshot.rs]: fixed

**Review by BusyJay — APPROVED** [review]

**Review by AndreMouche — APPROVED** [review]
